lib/widgets/editor/editor_toolbar.dart

```diff
@@ -0,0 +1,462 @@
+    var linesCount = 0;
+    while (lines.moveNext()) {
+      if (lines.isWithinSelection) {
+        if (!lines.current.startsWith(RegExp.escape(s))) {
+          lines.current = '$s${lines.current}';
+          linesCount++;
+        }
+      }
+    }
+
+    value = value.copyWith(
+      text: lines.text,
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset + s.length,
+        extentOffset: selection.extentOffset + s.length * linesCount,
+      ),
+    );
+  }
+
+  void reformat(Reformat Function(String selection) reformatter) {
+    final beg = beforeSelectionText;
+    final mid = selectionText;
+    final end = afterSelectionText;
+
+    final r = reformatter(mid);
+    value = value.copyWith(
+      text: '$beg${r.text}$end',
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset + r.selectionBeginningShift,
+        extentOffset: selection.extentOffset + r.selectionEndingShift,
+      ),
+    );
+  }
+
+  void reformatSimple(String text) =>
+      reformat((selection) => Reformat(text: text));
+}
+
+enum HeaderLevel {
+  h1(1),
+  h2(2),
+  h3(3),
+  h4(4),
+  h5(5),
+  h6(6);
+
+  const HeaderLevel(this.value);
+  final int value;
+}
+
+class Toolbar extends HookWidget {
+  final TextEditingController controller;
+  final String instanceHost;
+  final EditorToolbarStore store;
+  static const _height = 50.0;
+
+  Toolbar({
+    required this.controller,
+    required this.instanceHost,
+  }) : store = EditorToolbarStore(instanceHost);
+
+  @override
+  Widget build(BuildContext context) {
+    return MobxProvider.value(
+      value: store,
+      child: AsyncStoreListener(
+        asyncStore: store.imageUploadState,
+        child: Container(
+          height: _height,
+          width: double.infinity,
+          color: Theme.of(context).cardColor,
+          child: Material(
+            child: SingleChildScrollView(
+              scrollDirection: Axis.horizontal,
+              child: _ToolbarBody(
+                controller: controller,
+                instanceHost: instanceHost,
+              ),
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+
+  static Widget safeArea = const SizedBox(height: _height);
+}
+
+class _ToolbarBody extends HookWidget {
+  const _ToolbarBody({
+    required this.controller,
+    required this.instanceHost,
+  });
+
+  final TextEditingController controller;
+  final String instanceHost;
+
+  @override
+  Widget build(BuildContext context) {
+    final loggedInAction = useLoggedInAction(instanceHost);
+    return Row(
+      children: [
+        IconButton(
+          onPressed: () => controller.surround(
+              before: '**',
+              placeholder: L10n.of(context).insert_text_here_placeholder),
+          icon: const Icon(Icons.format_bold),
+          tooltip: L10n.of(context).editor_bold,
+        ),
+        IconButton(
+          onPressed: () => controller.surround(
+              before: '*',
```

Comment:
`'_'` is more popular for italics

Reply:
no

Reply:
lemmy-ui uses asterisks
![image](https://user-images.githubusercontent.com/10037914/186757147-425e45c5-78d6-4880-a8f4-db28c9c1ea15.png)